htlcswitch/pending_payment.go

```diff
@@ -0,0 +1,492 @@
+	<-chan PaymentResult, error) {
+
+	var paymentIDBytes [8]byte
+	binary.BigEndian.PutUint64(paymentIDBytes[:], pid)
+
+	resultChan := make(chan PaymentResult, 1)
+	err := store.db.View(func(tx *bbolt.Tx) error {
+		pendingPayments := tx.Bucket(pendingPaymentBucketKey)
+		if pendingPayments == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		bucket := pendingPayments.Bucket(paymentIDBytes[:])
+		if bucket == nil {
+			return ErrPaymentIDNotFound
+		}
+
+		// Check whether a result is already available.
+		var result PaymentResult = &PaymentSuccess{}
+		resultBytes := bucket.Get(successKey)
+		if resultBytes == nil {
+			// If no success result was available, check for
+			// failure.
+			result = &PaymentFailure{}
+			resultBytes = bucket.Get(failureKey)
+		}
+
+		if resultBytes == nil {
+			// Result not yet available, we will notify once a
+			// result is available.
+			// NOTE: we do this inside the db transaction, to avoid
+			// a race where the payment result is written to the db
+			// between this db tx and we add the resultChan to the
+			// map.
+			store.resultsMtx.Lock()
+			store.results[pid] = append(
+				store.results[pid], resultChan,
+			)
+			store.resultsMtx.Unlock()
+			return nil
+		}
+
+		// Decode the result we found, and send it on the result
+		// channel.
+		r := bytes.NewReader(resultBytes)
+		if err := result.Decode(r); err != nil {
+			return err
+		}
+
+		resultChan <- result
+		return nil
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return resultChan, nil
+}
+
+// fetchPendingPayments retrieves all payments from the store that are pending,
+// meaning no results are available yet, and can be re-forwarded on the
+// network.
+func (store *pendingPaymentStore) fetchPendingPayments() ([]*PendingPayment,
+	error) {
+
+	var ps []*PendingPayment
+	err := store.db.View(func(tx *bbolt.Tx) error {
+		pendingPayments := tx.Bucket(pendingPaymentBucketKey)
+		if pendingPayments == nil {
+			return nil
+		}
+
+		pendingPayments.ForEach(func(k, v []byte) error {
+			// Skip non-buckets.
+			if v != nil {
+				return nil
+			}
+
+			bucket := pendingPayments.Bucket(k)
+			if bucket == nil {
+				return fmt.Errorf("unable to create bucket")
+			}
+
+			pendingBytes := bucket.Get(pendingPaymentKey)
+			if pendingBytes == nil {
+				// If there's no serialied pending payment, the
+				// result is already available for this
+				// payment.
+				return nil
+			}
+
+			var err error
+			r := bytes.NewReader(pendingBytes)
+			p, err := deserializePendingPayment(r)
+			if err != nil {
+				return err
+			}
+
+			ps = append(ps, p)
+			return nil
+		})
+		return nil
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return ps, nil
+}
+
+// syncWithCircuitMap should be called at startup to delete the pending
+// payments which are no longer found in the circuit map. This removes pending
+// payments that was not successfully forwarded to the link before shutdown.
+func (store *pendingPaymentStore) syncWithCircuitMap(
```

Comment:
If we need to do this anyway, what is the use of persisting pending payments? Can't that list be reconstructed on startup from the circuit map?

So only persist the outcomes when they become available.